module/zfs/dsl_scan.c

```diff
@@ -607,12 +607,100 @@ dsl_scan_sync_state(dsl_scan_t *scn, dmu_tx_t *tx, state_sync_type_t sync_type)
 	}
 }
 
+typedef struct dsl_scan_params {
+	pool_scan_func_t dsp_func;
+	char *dsp_dsname;
+	pool_scan_flags_t dsp_flags;
+} dsl_scan_params_t;
+
+/* ARGSUSED */
+static int
+enqueue_cb(dsl_pool_t *dp, dsl_dataset_t *hds, void *arg)
+{
+	dsl_dataset_t *ds;
+	int err;
+	dsl_scan_t *scn = dp->dp_scan;
+
+	err = dsl_dataset_hold_obj(dp, hds->ds_object, FTAG, &ds);
+	if (err)
+		return (err);
+
+	while (dsl_dataset_phys(ds)->ds_prev_snap_obj != 0) {
+		dsl_dataset_t *prev;
+		err = dsl_dataset_hold_obj(dp,
+		    dsl_dataset_phys(ds)->ds_prev_snap_obj, FTAG, &prev);
+		if (err) {
+			dsl_dataset_rele(ds, FTAG);
+			return (err);
+		}
+
+		/*
+		 * If this is a clone, we don't need to worry about it for now.
+		 */
+		if ((scn->scn_phys.scn_flags & DSF_SCRUB_DATASET) == 0 &&
+		    dsl_dataset_phys(prev)->ds_next_snap_obj != ds->ds_object) {
+			dsl_dataset_rele(ds, FTAG);
+			dsl_dataset_rele(prev, FTAG);
+			return (0);
+		}
+		dsl_dataset_rele(ds, FTAG);
+		ds = prev;
+	}
+
+	scan_ds_queue_insert(scn, ds->ds_object,
+	    dsl_dataset_phys(ds)->ds_prev_snap_txg);
+	dsl_dataset_rele(ds, FTAG);
+	return (0);
+}
+
+/* ARGSUSED */
+static int
+enqueue_ds_cb(dsl_pool_t *dp, dsl_dataset_t *ds, void *arg)
+{
+	dsl_scan_t *scn = dp->dp_scan;
+
+	if (dsl_dataset_phys(ds)->ds_prev_snap_obj != 0) {
+		dsl_dataset_t *oldsnap;
+		uint64_t used, comp, uncomp;
+
+		VERIFY0(dsl_dataset_hold_obj(dp,
+		    dsl_dataset_phys(ds)->ds_prev_snap_obj, FTAG, &oldsnap));
+		VERIFY0(dsl_dataset_space_written(oldsnap, ds, &used,
+		    &comp, &uncomp));
+		scn->scn_phys.scn_to_examine += comp;
+		dsl_dataset_rele(oldsnap, FTAG);
+	} else {
+		scn->scn_phys.scn_to_examine +=
+		    dsl_dataset_phys(ds)->ds_compressed_bytes;
+	}
+
+	scan_ds_queue_insert(scn, ds->ds_object,
+	    dsl_dataset_phys(ds)->ds_prev_snap_txg);
```

Comment:
If you specify a filesystem (with or without `-r`), we scan the whole history of the filesystem, back to the origin?